// docusaurus-theme-openapi-docs/lib/theme/ApiDemoPanel/buildPostmanRequest.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _postmanCollection = _interopRequireDefault(require("@paloaltonetworks/postman-collection"));
var _cloneDeep = _interopRequireDefault(require("lodash/cloneDeep"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

function setQueryParams(postman, queryParams) {
  postman.url.query.clear();
  const qp = queryParams.map(param => {
    if (!param.value) {
      return undefined;
    }
    if (Array.isArray(param.value)) {
      return new _postmanCollection.default.QueryParam({
        key: param.name,
        value: param.value.join(",")
      });
    }

    // Parameter allows empty value: "/hello?extended"
    if (param.allowEmptyValue) {
      if (param.value === "true") {
        return new _postmanCollection.default.QueryParam({
          key: param.name,
          value: null
        });
      }
      return undefined;
    }
    return new _postmanCollection.default.QueryParam({
      key: param.name,
      value: param.value
    });
  }).filter(item => item !== undefined);
  if (qp.length > 0) {
    postman.addQueryParams(qp);
  }
}
function setPathParams(postman, queryParams) {
  const source = queryParams.map(param => {
    return new _postmanCollection.default.Variable({
      key: param.name,
      value: param.value || `:${param.name}`
    });
  });
  postman.url.variables.assimilate(source, false);
}
function buildCookie(cookieParams) {
  const cookies = cookieParams.map(param => {
    if (param.value && !Array.isArray(param.value)) {
      return new _postmanCollection.default.Cookie({
        // TODO: Is this right?
        path: "",
        domain: "",
        key: param.name,
        value: param.value
      });
    }
    return undefined;
  }).filter(item => item !== undefined);
  const list = new _postmanCollection.default.CookieList(null, cookies);
  return list.toString();
}
function setHeaders(postman, contentType, accept, cookie, headerParams, other) {
  postman.headers.clear();
  if (contentType) {
    postman.addHeader({
      key: "Content-Type",
      value: contentType
    });
  }
  if (accept) {
    postman.addHeader({
      key: "Accept",
      value: accept
    });
  }
  headerParams.forEach(param => {
    if (param.value && !Array.isArray(param.value)) {
      postman.addHeader({
        key: param.name,
        value: param.value
      });
    }
  });
  other.forEach(header => {
    postman.addHeader(header);
  });
  if (cookie) {
    postman.addHeader({
      key: "Cookie",
      value: cookie
    });
  }
}

// TODO: this is all a bit hacky
function setBody(clonedPostman, body) {
  var _body$content;
  if (clonedPostman.body === undefined) {
    return;
  }
  if (body.type === "empty") {
    clonedPostman.body = undefined;
    return;
  }
  if (body.type === "raw" && ((_body$content = body.content) === null || _body$content === void 0 ? void 0 : _body$content.type) === "file") {
    // treat it like file.
    clonedPostman.body.mode = "file";
    clonedPostman.body.file = {
      src: body.content.value.src
    };
    return;
  }
  switch (clonedPostman.body.mode) {
    case "raw":
      {
        var _body$content2, _body$content$value, _body$content3;
        // check file even though it should already be set from above
        if (body.type !== "raw" || ((_body$content2 = body.content) === null || _body$content2 === void 0 ? void 0 : _body$content2.type) === "file") {
          clonedPostman.body = undefined;
          return;
        }
        clonedPostman.body.raw = (_body$content$value = (_body$content3 = body.content) === null || _body$content3 === void 0 ? void 0 : _body$content3.value) !== null && _body$content$value !== void 0 ? _body$content$value : "";
        return;
      }
    case "formdata":
      {
        var _clonedPostman$body$f, _clonedPostman$body$f2;
        (_clonedPostman$body$f = clonedPostman.body.formdata) === null || _clonedPostman$body$f === void 0 ? void 0 : _clonedPostman$body$f.clear();
        if (body.type !== "form") {
          var _body$content4;
          // treat it like raw.
          clonedPostman.body.mode = "raw";
          clonedPostman.body.raw = `${(_body$content4 = body.content) === null || _body$content4 === void 0 ? void 0 : _body$content4.value}`;
          return;
        }
        const params = Object.entries(body.content).filter(entry => !!entry[1]).map(([key, content]) => {
          if (content.type === "file") {
            return new _postmanCollection.default.FormParam({
              key: key,
              ...content
            });
          }
          return new _postmanCollection.default.FormParam({
            key: key,
            value: content.value
          });
        });
        (_clonedPostman$body$f2 = clonedPostman.body.formdata) === null || _clonedPostman$body$f2 === void 0 ? void 0 : _clonedPostman$body$f2.assimilate(params, false);
        return;
      }
    case "urlencoded":
      {
        var _clonedPostman$body$u, _clonedPostman$body$u2;
        (_clonedPostman$body$u = clonedPostman.body.urlencoded) === null || _clonedPostman$body$u === void 0 ? void 0 : _clonedPostman$body$u.clear();
        if (body.type !== "form") {
          var _body$content5;
          // treat it like raw.
          clonedPostman.body.mode = "raw";
          clonedPostman.body.raw = `${(_body$content5 = body.content) === null || _body$content5 === void 0 ? void 0 : _body$content5.value}`;
          return;
        }
        const params = Object.entries(body.content).filter(entry => !!entry[1]).map(([key, content]) => {
          if (content.type !== "file" && content.value) {
            return new _postmanCollection.default.QueryParam({
              key: key,
              value: content.value
            });
          }
          return undefined;
        }).filter(item => item !== undefined);
        (_clonedPostman$body$u2 = clonedPostman.body.urlencoded) === null || _clonedPostman$body$u2 === void 0 ? void 0 : _clonedPostman$body$u2.assimilate(params, false);
        return;
      }
    default:
      return;
  }
}

// TODO: finish these types

function buildPostmanRequest(postman, {
  queryParams,
  pathParams,
  cookieParams,
  contentType,
  accept,
  headerParams,
  body,
  server,
  auth
}) {
  const clonedPostman = (0, _cloneDeep.default)(postman);
  clonedPostman.url.protocol = undefined;
  clonedPostman.url.host = [window.location.origin];
  if (server) {
    let url = server.url.replace(/\/$/, "");
    const variables = server.variables;
    if (variables) {
      Object.keys(variables).forEach(variable => {
        url = url.replace(`{${variable}}`, variables[variable].default);
      });
    }
    clonedPostman.url.host = [url];
  }
  setQueryParams(clonedPostman, queryParams);
  setPathParams(clonedPostman, pathParams);
  const cookie = buildCookie(cookieParams);
  let otherHeaders = [];
  let selectedAuth = [];
  if (auth.selected !== undefined) {
    selectedAuth = auth.options[auth.selected];
  }
  for (const a of selectedAuth) {
    // Bearer Auth
    if (a.type === "http" && a.scheme === "bearer") {
      const {
        token
      } = auth.data[a.key];
      if (token === undefined) {
        otherHeaders.push({
          key: "Authorization",
          value: "Bearer <TOKEN>"
        });
        continue;
      }
      otherHeaders.push({
        key: "Authorization",
        value: `Bearer ${token}`
      });
      continue;
    }
    if (a.type === "oauth2") {
      let token;
      if (auth.data[a.key]) {
        token = auth.data[a.key].token;
      }
      if (token === undefined) {
        otherHeaders.push({
          key: "Authorization",
          value: "Bearer <TOKEN>"
        });
        continue;
      }
      otherHeaders.push({
        key: "Authorization",
        value: `Bearer ${token}`
      });
      continue;
    }

    // Basic Auth
    if (a.type === "http" && a.scheme === "basic") {
      const {
        username,
        password
      } = auth.data[a.key];
      if (username === undefined || password === undefined) {
        continue;
      }
      otherHeaders.push({
        key: "Authorization",
        value: `Basic ${window.btoa(`${username}:${password}`)}`
      });
      continue;
    }

    // API Key
    if (a.type === "apiKey" && a.in === "header") {
      const {
        apiKey
      } = auth.data[a.key];
      if (apiKey === undefined) {
        otherHeaders.push({
          key: a.name,
          value: "<API_KEY_VALUE>"
        });
        continue;
      }
      otherHeaders.push({
        key: a.name,
        value: apiKey
      });
      continue;
    }
  }
  setHeaders(clonedPostman, contentType, accept, cookie, headerParams, otherHeaders);
  setBody(clonedPostman, body);
  return clonedPostman;
}
var _default = buildPostmanRequest;
exports.default = _default;
