// docusaurus-theme-openapi-docs/lib/theme/ApiDemoPanel/storage-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createStorage = createStorage;
exports.hashArray = hashArray;
var _cryptoJs = _interopRequireDefault(require("crypto-js"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

function hashArray(arr) {
  function hash(message) {
    return _cryptoJs.default.SHA1(message).toString();
  }
  const hashed = arr.map(item => hash(item));
  hashed.sort();
  const res = hashed.join();
  return hash(res);
}
function createStorage(persistance) {
  if (persistance === false) {
    return {
      getItem: () => null,
      setItem: () => {},
      clear: () => {},
      key: () => null,
      removeItem: () => {},
      length: 0
    };
  }
  if (persistance === "sessionStorage") {
    return sessionStorage;
  }
  return localStorage;
}
