// docusaurus-theme-openapi-docs/lib/theme/ApiDemoPanel/ParamOptions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _toolkit = require("@reduxjs/toolkit");
var _FormItem = _interopRequireDefault(require("@theme/ApiDemoPanel/FormItem"));
var _FormMultiSelect = _interopRequireDefault(require("@theme/ApiDemoPanel/FormMultiSelect"));
var _FormSelect = _interopRequireDefault(require("@theme/ApiDemoPanel/FormSelect"));
var _FormTextInput = _interopRequireDefault(require("@theme/ApiDemoPanel/FormTextInput"));
var _hooks = require("@theme/ApiItem/hooks");
var _slice = require("./slice");
var _stylesModule = _interopRequireDefault(require("./styles.module.css"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

function ParamOption({
  param
}) {
  var _param$schema, _param$schema$items, _param$schema2, _param$schema3, _param$schema4;
  if (((_param$schema = param.schema) === null || _param$schema === void 0 ? void 0 : _param$schema.type) === "array" && (_param$schema$items = param.schema.items) !== null && _param$schema$items !== void 0 && _param$schema$items.enum) {
    return <ParamMultiSelectFormItem param={param} />;
  }
  if (((_param$schema2 = param.schema) === null || _param$schema2 === void 0 ? void 0 : _param$schema2.type) === "array") {
    return <ParamArrayFormItem param={param} />;
  }
  if ((_param$schema3 = param.schema) !== null && _param$schema3 !== void 0 && _param$schema3.enum) {
    return <ParamSelectFormItem param={param} />;
  }
  if (((_param$schema4 = param.schema) === null || _param$schema4 === void 0 ? void 0 : _param$schema4.type) === "boolean") {
    return <ParamBooleanFormItem param={param} />;
  }

  // integer, number, string, int32, int64, float, double, object, byte, binary,
  // date-time, date, password
  return <ParamTextFormItem param={param} />;
}
function ParamOptionWrapper({
  param
}) {
  return <_FormItem.default label={param.name} type={param.in} required={param.required}>
      <ParamOption param={param} />
    </_FormItem.default>;
}
function ParamOptions() {
  const [showOptional, setShowOptional] = (0, _react.useState)(false);
  const pathParams = (0, _hooks.useTypedSelector)(state => state.params.path);
  const queryParams = (0, _hooks.useTypedSelector)(state => state.params.query);
  const cookieParams = (0, _hooks.useTypedSelector)(state => state.params.cookie);
  const headerParams = (0, _hooks.useTypedSelector)(state => state.params.header);
  const allParams = [...pathParams, ...queryParams, ...cookieParams, ...headerParams];
  const requiredParams = allParams.filter(p => p.required);
  const optionalParams = allParams.filter(p => !p.required);
  return <>
      {/* Required Parameters */}
      {requiredParams.map(param => <ParamOptionWrapper key={`${param.in}-${param.name}`} param={param} />)}

      {/* Optional Parameters */}
      {optionalParams.length > 0 && <>
          <button className={_stylesModule.default.showMoreButton} onClick={() => setShowOptional(prev => !prev)}>
            <span style={{
          width: "1.5em",
          display: "inline-block",
          textAlign: "center"
        }}>
              <span className={showOptional ? _stylesModule.default.plusExpanded : _stylesModule.default.plus}>
                <div>
                  <svg style={{
                fill: "currentColor",
                width: "10px",
                height: "10px"
              }} height="16" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg">
                    <path d="M9 7h6a1 1 0 0 1 0 2H9v6a1 1 0 0 1-2 0V9H1a1 1 0 1 1 0-2h6V1a1 1 0 1 1 2 0z" fillRule="evenodd"></path>
                  </svg>
                </div>
              </span>
            </span>
            {showOptional ? "Hide optional parameters" : "Show optional parameters"}
          </button>

          <div className={showOptional ? _stylesModule.default.showOptions : _stylesModule.default.hideOptions}>
            {optionalParams.map(param => <ParamOptionWrapper key={`${param.in}-${param.name}`} param={param} />)}
          </div>
        </>}
    </>;
}
function ArrayItem({
  param,
  onChange
}) {
  var _param$schema5, _param$schema5$items;
  if (((_param$schema5 = param.schema) === null || _param$schema5 === void 0 ? void 0 : (_param$schema5$items = _param$schema5.items) === null || _param$schema5$items === void 0 ? void 0 : _param$schema5$items.type) === "boolean") {
    return <_FormSelect.default options={["---", "true", "false"]} onChange={e => {
      const val = e.target.value;
      onChange(val === "---" ? undefined : val);
    }} />;
  }
  return <_FormTextInput.default placeholder={param.description || param.name} onChange={e => {
    onChange(e.target.value);
  }} />;
}
function ParamArrayFormItem({
  param
}) {
  const [items, setItems] = (0, _react.useState)([]);
  const dispatch = (0, _hooks.useTypedDispatch)();
  function handleAddItem() {
    setItems(i => [...i, {
      id: (0, _toolkit.nanoid)()
    }]);
  }
  (0, _react.useEffect)(() => {
    const values = items.map(item => item.value).filter(item => !!item);
    dispatch((0, _slice.setParam)({
      ...param,
      value: values.length > 0 ? values : undefined
    }));
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [items]);
  function handleDeleteItem(itemToDelete) {
    return () => {
      const newItems = items.filter(i => i.id !== itemToDelete.id);
      setItems(newItems);
    };
  }
  function handleChangeItem(itemToUpdate) {
    return value => {
      const newItems = items.map(i => {
        if (i.id === itemToUpdate.id) {
          return {
            ...i,
            value: value
          };
        }
        return i;
      });
      setItems(newItems);
    };
  }
  return <>
      {items.map(item => <div key={item.id} style={{
      display: "flex"
    }}>
          <ArrayItem param={param} onChange={handleChangeItem(item)} />
          <button className={_stylesModule.default.buttonDelete} onClick={handleDeleteItem(item)}>
            <svg focusable="false" preserveAspectRatio="xMidYMid meet" xmlns="http://www.w3.org/2000/svg" fill="currentColor" width="16" height="16" viewBox="0 0 32 32" aria-hidden="true">
              <path d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"></path>
              <title>Delete</title>
            </svg>
          </button>
        </div>)}
      <button className={_stylesModule.default.buttonThin} onClick={handleAddItem}>
        Add item
      </button>
    </>;
}
function ParamSelectFormItem({
  param
}) {
  var _param$schema$enum, _param$schema6;
  const dispatch = (0, _hooks.useTypedDispatch)();
  const options = (_param$schema$enum = (_param$schema6 = param.schema) === null || _param$schema6 === void 0 ? void 0 : _param$schema6.enum) !== null && _param$schema$enum !== void 0 ? _param$schema$enum : [];
  return <_FormSelect.default options={["---", ...options]} onChange={e => {
    const val = e.target.value;
    dispatch((0, _slice.setParam)({
      ...param,
      value: val === "---" ? undefined : val
    }));
  }} />;
}
function ParamBooleanFormItem({
  param
}) {
  const dispatch = (0, _hooks.useTypedDispatch)();
  return <_FormSelect.default options={["---", "true", "false"]} onChange={e => {
    const val = e.target.value;
    dispatch((0, _slice.setParam)({
      ...param,
      value: val === "---" ? undefined : val
    }));
  }} />;
}
function ParamMultiSelectFormItem({
  param
}) {
  var _param$schema$items$e, _param$schema7, _param$schema7$items;
  const dispatch = (0, _hooks.useTypedDispatch)();
  const options = (_param$schema$items$e = (_param$schema7 = param.schema) === null || _param$schema7 === void 0 ? void 0 : (_param$schema7$items = _param$schema7.items) === null || _param$schema7$items === void 0 ? void 0 : _param$schema7$items.enum) !== null && _param$schema$items$e !== void 0 ? _param$schema$items$e : [];
  return <_FormMultiSelect.default options={options} onChange={e => {
    const values = Array.prototype.filter.call(e.target.options, o => o.selected).map(o => o.value);
    dispatch((0, _slice.setParam)({
      ...param,
      value: values.length > 0 ? values : undefined
    }));
  }} />;
}
function ParamTextFormItem({
  param
}) {
  const dispatch = (0, _hooks.useTypedDispatch)();
  return <_FormTextInput.default placeholder={param.description || param.name} onChange={e => dispatch((0, _slice.setParam)({
    ...param,
    value: param.in === "path" || param.in === "query" ? e.target.value.replace(/\s/g, "%20") : e.target.value
  }))} />;
}
var _default = ParamOptions;
exports.default = _default;
