// docusaurus-theme-openapi-docs/lib-next/theme/ApiDemoPanel/Body/slice.js
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

import { createSlice } from "@reduxjs/toolkit";
const initialState = {};
export const slice = createSlice({
  name: "body",
  initialState,
  reducers: {
    clearRawBody: (_state) => {
      return {
        type: "empty",
      };
    },
    setStringRawBody: (_state, action) => {
      return {
        type: "raw",
        content: {
          type: "string",
          value: action.payload,
        },
      };
    },
    setFileRawBody: (_state, action) => {
      return {
        type: "raw",
        content: {
          type: "file",
          value: action.payload,
        },
      };
    },
    clearFormBodyKey: (state, action) => {
      if (state?.type === "form") {
        delete state.content[action.payload];
      }
    },
    setStringFormBody: (state, action) => {
      if (state?.type !== "form") {
        return {
          type: "form",
          content: {
            [action.payload.key]: {
              type: "string",
              value: action.payload.value,
            },
          },
        };
      }
      state.content[action.payload.key] = {
        type: "string",
        value: action.payload.value,
      };
      return state;
    },
    setFileFormBody: (state, action) => {
      if (state?.type !== "form") {
        return {
          type: "form",
          content: {
            [action.payload.key]: {
              type: "file",
              value: action.payload.value,
            },
          },
        };
      }
      state.content[action.payload.key] = {
        type: "file",
        value: action.payload.value,
      };
      return state;
    },
  },
});
export const {
  clearRawBody,
  setStringRawBody,
  setFileRawBody,
  clearFormBodyKey,
  setStringFormBody,
  setFileFormBody,
} = slice.actions;
export default slice.reducer;
