// docusaurus-theme-openapi-docs/lib/theme/ApiDemoPanel/Body/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _json2xml = _interopRequireDefault(require("@theme/ApiDemoPanel/Body/json2xml"));
var _ContentType = _interopRequireDefault(require("@theme/ApiDemoPanel/ContentType"));
var _FormFileUpload = _interopRequireDefault(require("@theme/ApiDemoPanel/FormFileUpload"));
var _FormItem = _interopRequireDefault(require("@theme/ApiDemoPanel/FormItem"));
var _FormSelect = _interopRequireDefault(require("@theme/ApiDemoPanel/FormSelect"));
var _FormTextInput = _interopRequireDefault(require("@theme/ApiDemoPanel/FormTextInput"));
var _LiveEditor = _interopRequireDefault(require("@theme/ApiDemoPanel/LiveEditor"));
var _hooks = require("@theme/ApiItem/hooks");
var _SchemaTabs = _interopRequireDefault(require("@theme/SchemaTabs"));
var _TabItem = _interopRequireDefault(require("@theme/TabItem"));
var _xmlFormatter = _interopRequireDefault(require("xml-formatter"));
var _slice = require("./slice");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

function BodyWrap({
  requestBodyMetadata,
  jsonRequestBodyExample
}) {
  const contentType = (0, _hooks.useTypedSelector)(state => state.contentType.value);

  // NOTE: We used to check if body was required, but opted to always show the request body
  // to reduce confusion, see: https://github.com/cloud-annotations/docusaurus-openapi/issues/145

  // No body
  if (contentType === undefined) {
    return null;
  }
  return <>
      <_ContentType.default />
      <Body requestBodyMetadata={requestBodyMetadata} jsonRequestBodyExample={jsonRequestBodyExample} />
    </>;
}
function Body({
  requestBodyMetadata,
  jsonRequestBodyExample
}) {
  var _requestBodyMetadata$, _requestBodyMetadata$2, _requestBodyMetadata$3, _requestBodyMetadata$4, _requestBodyMetadata$5, _requestBodyMetadata$6;
  const contentType = (0, _hooks.useTypedSelector)(state => state.contentType.value);
  const required = requestBodyMetadata === null || requestBodyMetadata === void 0 ? void 0 : requestBodyMetadata.required;
  const dispatch = (0, _hooks.useTypedDispatch)();

  // Lot's of possible content-types:
  // - application/json
  // - application/xml
  // - text/plain
  // - text/css
  // - text/html
  // - text/javascript
  // - application/javascript
  // - multipart/form-data
  // - application/x-www-form-urlencoded
  // - image/svg+xml;charset=US-ASCII

  // Show editor:
  // - application/json
  // - application/xml
  // - */*

  // Show form:
  // - multipart/form-data
  // - application/x-www-form-urlencoded

  const schema = requestBodyMetadata === null || requestBodyMetadata === void 0 ? void 0 : (_requestBodyMetadata$ = requestBodyMetadata.content) === null || _requestBodyMetadata$ === void 0 ? void 0 : (_requestBodyMetadata$2 = _requestBodyMetadata$[contentType]) === null || _requestBodyMetadata$2 === void 0 ? void 0 : _requestBodyMetadata$2.schema;
  const example = requestBodyMetadata === null || requestBodyMetadata === void 0 ? void 0 : (_requestBodyMetadata$3 = requestBodyMetadata.content) === null || _requestBodyMetadata$3 === void 0 ? void 0 : (_requestBodyMetadata$4 = _requestBodyMetadata$3[contentType]) === null || _requestBodyMetadata$4 === void 0 ? void 0 : _requestBodyMetadata$4.example;
  const examples = requestBodyMetadata === null || requestBodyMetadata === void 0 ? void 0 : (_requestBodyMetadata$5 = requestBodyMetadata.content) === null || _requestBodyMetadata$5 === void 0 ? void 0 : (_requestBodyMetadata$6 = _requestBodyMetadata$5[contentType]) === null || _requestBodyMetadata$6 === void 0 ? void 0 : _requestBodyMetadata$6.examples;
  if ((schema === null || schema === void 0 ? void 0 : schema.format) === "binary") {
    return <_FormItem.default label="Body" required={required}>
        <_FormFileUpload.default placeholder={schema.description || "Body"} onChange={file => {
        if (file === undefined) {
          dispatch((0, _slice.clearRawBody)());
          return;
        }
        dispatch((0, _slice.setFileRawBody)({
          src: `/path/to/${file.name}`,
          content: file
        }));
      }} />
      </_FormItem.default>;
  }
  if ((contentType === "multipart/form-data" || contentType === "application/x-www-form-urlencoded") && (schema === null || schema === void 0 ? void 0 : schema.type) === "object") {
    var _schema$properties;
    return <_FormItem.default label="Body" required={required}>
        <div style={{
        marginTop: "calc(var(--ifm-pre-padding) / 2)",
        borderRadius: "4px",
        padding: "var(--ifm-pre-padding)",
        border: "1px solid var(--openapi-monaco-border-color)"
      }}>
          {Object.entries((_schema$properties = schema.properties) !== null && _schema$properties !== void 0 ? _schema$properties : {}).map(([key, val]) => {
          if (val.format === "binary") {
            return <_FormItem.default key={key} label={key} required={Array.isArray(schema.required) && schema.required.includes(key)}>
                  <_FormFileUpload.default placeholder={val.description || key} onChange={file => {
                if (file === undefined) {
                  dispatch((0, _slice.clearFormBodyKey)(key));
                  return;
                }
                dispatch((0, _slice.setFileFormBody)({
                  key: key,
                  value: {
                    src: `/path/to/${file.name}`,
                    content: file
                  }
                }));
              }} />
                </_FormItem.default>;
          }
          if (val.enum) {
            return <_FormItem.default key={key} label={key} required={Array.isArray(schema.required) && schema.required.includes(key)}>
                  <_FormSelect.default options={["---", ...val.enum]} onChange={e => {
                const val = e.target.value;
                if (val === "---") {
                  dispatch((0, _slice.clearFormBodyKey)(key));
                } else {
                  dispatch((0, _slice.setStringFormBody)({
                    key: key,
                    value: val
                  }));
                }
              }} />
                </_FormItem.default>;
          }
          // TODO: support all the other types.
          return <_FormItem.default key={key} label={key} required={Array.isArray(schema.required) && schema.required.includes(key)}>
                <_FormTextInput.default placeholder={val.description || key} onChange={e => {
              dispatch((0, _slice.setStringFormBody)({
                key: key,
                value: e.target.value
              }));
            }} />
              </_FormItem.default>;
        })}
        </div>
      </_FormItem.default>;
  }
  let language = "plaintext";
  let defaultBody = ""; //"body content";
  let exampleBody;
  let examplesBodies = [];
  if (contentType === "application/json" || contentType.endsWith("+json")) {
    if (jsonRequestBodyExample) {
      defaultBody = JSON.stringify(jsonRequestBodyExample, null, 2);
    }
    if (example) {
      exampleBody = JSON.stringify(example, null, 2);
    }
    if (examples) {
      for (const [key, example] of Object.entries(examples)) {
        examplesBodies.push({
          label: key,
          body: JSON.stringify(example.value, null, 2),
          summary: example.summary
        });
      }
    }
    language = "json";
  }
  if (contentType === "application/xml" || contentType.endsWith("+xml")) {
    if (jsonRequestBodyExample) {
      try {
        defaultBody = (0, _xmlFormatter.default)((0, _json2xml.default)(jsonRequestBodyExample, ""), {
          indentation: "  ",
          lineSeparator: "\n",
          collapseContent: true
        });
      } catch {
        defaultBody = (0, _json2xml.default)(jsonRequestBodyExample);
      }
    }
    if (example) {
      try {
        exampleBody = (0, _xmlFormatter.default)((0, _json2xml.default)(example, ""), {
          indentation: "  ",
          lineSeparator: "\n",
          collapseContent: true
        });
      } catch {
        exampleBody = (0, _json2xml.default)(example);
      }
    }
    if (examples) {
      for (const [key, example] of Object.entries(examples)) {
        let formattedXmlBody;
        try {
          formattedXmlBody = (0, _xmlFormatter.default)(example.value, {
            indentation: "  ",
            lineSeparator: "\n",
            collapseContent: true
          });
        } catch {
          formattedXmlBody = example.value;
        }
        examplesBodies.push({
          label: key,
          body: formattedXmlBody,
          summary: example.summary
        });
      }
    }
    language = "xml";
  }
  if (exampleBody) {
    return <_FormItem.default label="Body" required={required}>
        <_SchemaTabs.default lazy>
          <_TabItem.default label="Default" value="default" default>
            <_LiveEditor.default action={dispatch} language={language}>
              {defaultBody}
            </_LiveEditor.default>
          </_TabItem.default>
          <_TabItem.default label="Example" value="example">
            {exampleBody && <_LiveEditor.default action={dispatch} language={language}>
                {exampleBody}
              </_LiveEditor.default>}
          </_TabItem.default>
        </_SchemaTabs.default>
      </_FormItem.default>;
  }
  if (examplesBodies && examplesBodies.length > 0) {
    return <_FormItem.default label="Body" required={required}>
        <_SchemaTabs.default lazy>
          <_TabItem.default label="Default" value="default" default>
            <_LiveEditor.default action={dispatch} language={language}>
              {defaultBody}
            </_LiveEditor.default>
          </_TabItem.default>
          {examplesBodies.map(example => {
          return <_TabItem.default label={example.label} value={example.label} key={example.label}>
                {example.summary && <p>{example.summary}</p>}
                {example.body && <_LiveEditor.default action={dispatch} language={language}>
                    {example.body}
                  </_LiveEditor.default>}
              </_TabItem.default>;
        })}
        </_SchemaTabs.default>
      </_FormItem.default>;
  }
  return <_FormItem.default label="Body" required={required}>
      <_LiveEditor.default action={dispatch} language={language}>
        {defaultBody}
      </_LiveEditor.default>
    </_FormItem.default>;
}
var _default = BodyWrap;
exports.default = _default;
