// docusaurus-theme-openapi-docs/lib/theme/ApiDemoPanel/Execute/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _buildPostmanRequest = _interopRequireDefault(require("@theme/ApiDemoPanel/buildPostmanRequest"));
var _slice = require("@theme/ApiDemoPanel/Response/slice");
var _hooks = require("@theme/ApiItem/hooks");
var _reactModal = _interopRequireDefault(require("react-modal"));
var _makeRequest = _interopRequireDefault(require("./makeRequest"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

function validateRequest(params) {
  for (let paramList of Object.values(params)) {
    for (let param of paramList) {
      if (param.required && !param.value) {
        return false;
      }
    }
  }
  return true;
}
function Execute({
  postman,
  proxy
}) {
  const pathParams = (0, _hooks.useTypedSelector)(state => state.params.path);
  const queryParams = (0, _hooks.useTypedSelector)(state => state.params.query);
  const cookieParams = (0, _hooks.useTypedSelector)(state => state.params.cookie);
  const headerParams = (0, _hooks.useTypedSelector)(state => state.params.header);
  const contentType = (0, _hooks.useTypedSelector)(state => state.contentType.value);
  const body = (0, _hooks.useTypedSelector)(state => state.body);
  const accept = (0, _hooks.useTypedSelector)(state => state.accept.value);
  const server = (0, _hooks.useTypedSelector)(state => state.server.value);
  const params = (0, _hooks.useTypedSelector)(state => state.params);
  const auth = (0, _hooks.useTypedSelector)(state => state.auth);
  const isValidRequest = validateRequest(params);
  const dispatch = (0, _hooks.useTypedDispatch)();
  const postmanRequest = (0, _buildPostmanRequest.default)(postman, {
    queryParams,
    pathParams,
    cookieParams,
    contentType,
    accept,
    headerParams,
    body,
    server,
    auth
  });
  const delay = ms => new Promise(resolve => setTimeout(resolve, ms));
  function openModal() {
    setIsOpen(true);
  }
  function closeModal() {
    setIsOpen(false);
  }
  function acceptAgreement() {
    setIsOpen(false);
    setAgreementAccepted(true);
    sessionStorage.setItem("agreement-ack", "true");
  }
  const [modalIsOpen, setIsOpen] = _react.default.useState(false);
  // Set the following as default value to persist to session and enable modal
  // sessionStorage.getItem("agreement-ack") === "true"
  const [agreementAccepted, setAgreementAccepted] = _react.default.useState(true);
  const customStyles = {
    overlay: {
      backdropFilter: "blur(10px)",
      backgroundColor: "transparent"
    },
    content: {
      top: "50%",
      left: "50%",
      right: "auto",
      bottom: "auto",
      marginRight: "-50%",
      border: "none",
      padding: "none",
      borderRadius: "var(--openapi-card-border-radius)",
      background: "var(--ifm-card-background-color)",
      transform: "translate(-50%, -50%)",
      maxWidth: "550px"
    }
  };
  if (agreementAccepted) {
    return <button className="button button--sm button--secondary" disabled={!isValidRequest} style={!isValidRequest ? {
      pointerEvents: "all"
    } : {}} onClick={async () => {
      dispatch((0, _slice.setResponse)("Fetching..."));
      try {
        await delay(1200);
        const res = await (0, _makeRequest.default)(postmanRequest, proxy, body);
        dispatch((0, _slice.setResponse)(res));
      } catch (e) {
        console.log(e);
        dispatch((0, _slice.setResponse)("Connection failed"));
      }
    }}>
        Send API Request
      </button>;
  } else {
    return <_react.default.Fragment>
        <button className="button button--sm button--secondary" onClick={openModal}>
          Send API Request
        </button>
        <_reactModal.default isOpen={modalIsOpen} onRequestClose={closeModal} style={customStyles} contentLabel="Terms of Use">
          <form>
            <div className="card">
              <div className="card__header">
                <h2>Terms of Use</h2>
                <hr></hr>
              </div>
              <div className="card__body">
                <p>
                  By accepting this agreement the end user acknowledges the
                  risks of performing authenticated and non-authenticated API
                  requests from the browser.
                </p>
                <p>
                  The end user also accepts the responsibility of safeguarding
                  API credentials and any potentially sensitive data returned by
                  the API.
                </p>
                <br></br>
              </div>
              <div className="card__footer">
                <div className="button-group button-group--block">
                  <button className="button button--sm button--outline button--success" onClick={acceptAgreement}>
                    AGREE
                  </button>
                  <button className="button button--sm button--outline button--danger" onClick={closeModal}>
                    DISAGREE
                  </button>
                </div>
              </div>
            </div>
          </form>
        </_reactModal.default>
      </_react.default.Fragment>;
  }
}
var _default = Execute;
exports.default = _default;
