// docusaurus-theme-openapi-docs/lib/theme/ApiDemoPanel/Authorization/slice.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createAuth = createAuth;
exports.slice = exports.setSelectedAuth = exports.setAuthData = exports.default = void 0;
var _toolkit = require("@reduxjs/toolkit");
var _storageUtils = require("@theme/ApiDemoPanel/storage-utils");
var _authTypes = require("./auth-types");
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

// The global definitions
// "securitySchemes": {
//   "BearerAuth": { "type": "http", "scheme": "BeAreR" },
//   "BasicAuth": { "type": "http", "scheme": "basic" }
// },

// The operation level requirements
// "security": [
//   { "BearerAuth": [] },
//   { "BearerAuth": [], "BasicAuth": [] }
// ],

// SLICE_STATE
// data:
//   BearerAuth:
//     token=xxx
//   BasicAuth:
//     username=xxx
//     password=xxx
//
// options:
//    "BearerAuth": [{ key: "BearerAuth", scopes: [], ...rest }]
//    "BearerAuth and BasicAuth": [{ key: "BearerAuth", scopes: [], ...rest }, { key: "BasicAuth", scopes: [], ...rest }]
//
// selected: "BearerAuth and BasicAuth"

// LOCAL_STORAGE
// hash(SLICE_STATE.options) -> "BearerAuth and BasicAuth"
// BearerAuth                -> { token: xxx }
// BasicAuth                 -> { username: xxx, password: xxx }

function createAuth({
  security,
  securitySchemes,
  options: opts
}) {
  var _persisted;
  const storage = (0, _storageUtils.createStorage)("sessionStorage");
  let data = {};
  let options = {};
  for (const option of security !== null && security !== void 0 ? security : []) {
    const id = Object.keys(option).join(" and ");
    for (const [schemeID, scopes] of Object.entries(option)) {
      const scheme = securitySchemes === null || securitySchemes === void 0 ? void 0 : securitySchemes[schemeID];
      if (scheme) {
        if (options[id] === undefined) {
          options[id] = [];
        }
        const dataKeys = (0, _authTypes.getAuthDataKeys)(scheme);
        for (const key of dataKeys) {
          if (data[schemeID] === undefined) {
            data[schemeID] = {};
          }
          let persisted = undefined;
          try {
            var _storage$getItem;
            persisted = JSON.parse((_storage$getItem = storage.getItem(schemeID)) !== null && _storage$getItem !== void 0 ? _storage$getItem : "")[key];
          } catch {}
          data[schemeID][key] = persisted;
        }
        options[id].push({
          ...scheme,
          key: schemeID,
          scopes
        });
      }
    }
  }
  let persisted = undefined;
  try {
    var _storage$getItem2;
    persisted = (_storage$getItem2 = storage.getItem((0, _storageUtils.hashArray)(Object.keys(options)))) !== null && _storage$getItem2 !== void 0 ? _storage$getItem2 : undefined;
  } catch {}
  return {
    data,
    options,
    selected: (_persisted = persisted) !== null && _persisted !== void 0 ? _persisted : Object.keys(options)[0]
  };
}
const initialState = {};
const slice = (0, _toolkit.createSlice)({
  name: "auth",
  initialState,
  reducers: {
    setAuthData: (state, action) => {
      const {
        scheme,
        key,
        value
      } = action.payload;
      state.data[scheme][key] = value;
    },
    setSelectedAuth: (state, action) => {
      state.selected = action.payload;
    }
  }
});
exports.slice = slice;
const {
  setAuthData,
  setSelectedAuth
} = slice.actions;
exports.setSelectedAuth = setSelectedAuth;
exports.setAuthData = setAuthData;
var _default = slice.reducer;
exports.default = _default;
