// docusaurus-protobuffet/dist/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var pluginOptionDefaults = {
    protoDocsPath: './protodocs',
    sidebarPath: './sidebarsProtodocs.js',
};
function preset(context, options) {
    var _a;
    var pluginOptions = __assign(__assign({}, pluginOptionDefaults), options.protobuffet);
    var docSidebarPath = ((_a = options.docs) === null || _a === void 0 ? void 0 : _a.sidebarPath) || pluginOptions.sidebarPath;
    var config = {
        plugins: [
            [
                require.resolve('docusaurus-protobuffet-plugin'),
                pluginOptions,
            ],
            [
                '@docusaurus/plugin-content-docs',
                {
                    id: 'protodocs',
                    path: 'protodocs',
                    routeBasePath: 'protodocs',
                    sidebarPath: docSidebarPath,
                },
            ],
        ],
    };
    return config;
}
exports.default = preset;
