// docusaurus-theme-openapi-docs/lib/theme/ApiItem/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ApiItem;
var _react = _interopRequireDefault(require("react"));
var _BrowserOnly = _interopRequireDefault(require("@docusaurus/BrowserOnly"));
var _ExecutionEnvironment = _interopRequireDefault(require("@docusaurus/ExecutionEnvironment"));
var _themeCommon = require("@docusaurus/theme-common");
var _useDocusaurusContext = _interopRequireDefault(require("@docusaurus/useDocusaurusContext"));
var _useIsBrowser = _interopRequireDefault(require("@docusaurus/useIsBrowser"));
var _slice = require("@theme/ApiDemoPanel/Authorization/slice");
var _persistanceMiddleware = require("@theme/ApiDemoPanel/persistanceMiddleware");
var _Layout = _interopRequireDefault(require("@theme/ApiItem/Layout"));
var _Metadata = _interopRequireDefault(require("@theme/DocItem/Metadata"));
var _clsx = _interopRequireDefault(require("clsx"));
var _reactRedux = require("react-redux");
var _store = require("./store");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

const {
  DocProvider
} = require("@docusaurus/theme-common/internal");
let ApiDemoPanel = _ => <div />;
if (_ExecutionEnvironment.default.canUseDOM) {
  ApiDemoPanel = require("@theme/ApiDemoPanel").default;
}
function ApiItem(props) {
  const docHtmlClassName = `docs-doc-id-${props.content.metadata.unversionedId}`;
  const MDXComponent = props.content;
  const {
    frontMatter
  } = MDXComponent;
  const {
    info_path: infoPath
  } = frontMatter;
  const {
    api
  } = frontMatter;
  const {
    siteConfig
  } = (0, _useDocusaurusContext.default)();
  const themeConfig = siteConfig.themeConfig;
  const options = themeConfig.api;
  const isBrowser = (0, _useIsBrowser.default)();

  // Regex for 2XX status
  const statusRegex = new RegExp("(20[0-9]|2[1-9][0-9])");

  // Define store2
  let store2 = {};
  const persistanceMiddleware = (0, _persistanceMiddleware.createPersistanceMiddleware)(options);

  // Init store for SSR
  if (!isBrowser) {
    store2 = (0, _store.createStoreWithoutState)({}, [persistanceMiddleware]);
  }

  // Init store for CSR to hydrate components
  if (isBrowser) {
    var _api$requestBody$cont, _api$requestBody, _api$servers, _api$parameters, _window, _ref;
    // Create list of only 2XX response content types to create request samples from
    let acceptArray = [];
    for (const [code, content] of Object.entries((_api$responses = api === null || api === void 0 ? void 0 : api.responses) !== null && _api$responses !== void 0 ? _api$responses : [])) {
      var _api$responses;
      if (statusRegex.test(code)) {
        var _content$content;
        acceptArray.push(Object.keys((_content$content = content.content) !== null && _content$content !== void 0 ? _content$content : {}));
      }
    }
    acceptArray = acceptArray.flat();
    const content = (_api$requestBody$cont = api === null || api === void 0 ? void 0 : (_api$requestBody = api.requestBody) === null || _api$requestBody === void 0 ? void 0 : _api$requestBody.content) !== null && _api$requestBody$cont !== void 0 ? _api$requestBody$cont : {};
    const contentTypeArray = Object.keys(content);
    const servers = (_api$servers = api === null || api === void 0 ? void 0 : api.servers) !== null && _api$servers !== void 0 ? _api$servers : [];
    const params = {
      path: [],
      query: [],
      header: [],
      cookie: []
    };
    api === null || api === void 0 ? void 0 : (_api$parameters = api.parameters) === null || _api$parameters === void 0 ? void 0 : _api$parameters.forEach(param => {
      const paramType = param.in;
      const paramsArray = params[paramType];
      paramsArray.push(param);
    });
    const auth = (0, _slice.createAuth)({
      security: api === null || api === void 0 ? void 0 : api.security,
      securitySchemes: api === null || api === void 0 ? void 0 : api.securitySchemes,
      options
    });
    // TODO: determine way to rehydrate without flashing
    // const acceptValue = window?.sessionStorage.getItem("accept");
    // const contentTypeValue = window?.sessionStorage.getItem("contentType");
    const server = (_window = window) === null || _window === void 0 ? void 0 : _window.sessionStorage.getItem("server");
    const serverObject = (_ref = JSON.parse(server)) !== null && _ref !== void 0 ? _ref : {};
    store2 = (0, _store.createStoreWithState)({
      accept: {
        value: acceptArray[0],
        options: acceptArray
      },
      contentType: {
        value: contentTypeArray[0],
        options: contentTypeArray
      },
      server: {
        value: serverObject.url ? serverObject : undefined,
        options: servers
      },
      response: {
        value: undefined
      },
      body: {
        type: "empty"
      },
      params,
      auth
    }, [persistanceMiddleware]);
  }
  if (api) {
    return <DocProvider content={props.content}>
        <_themeCommon.HtmlClassNameProvider className={docHtmlClassName}>
          <_Metadata.default />
          <_Layout.default>
            <_reactRedux.Provider store={store2}>
              <div className={(0, _clsx.default)("row", "theme-api-markdown")}>
                <div className="col col--7">
                  <MDXComponent />
                </div>
                <div className="col col--5">
                  <_BrowserOnly.default fallback={<div>Loading...</div>}>
                    {() => {
                    return <ApiDemoPanel item={api} infoPath={infoPath} />;
                  }}
                  </_BrowserOnly.default>
                </div>
              </div>
            </_reactRedux.Provider>
          </_Layout.default>
        </_themeCommon.HtmlClassNameProvider>
      </DocProvider>;
  }

  // Non-API docs
  return <DocProvider content={props.content}>
      <_themeCommon.HtmlClassNameProvider className={docHtmlClassName}>
        <_Metadata.default />
        <_Layout.default>
          <div className="row">
            <div className="col col--12">
              <MDXComponent />
            </div>
          </div>
        </_Layout.default>
      </_themeCommon.HtmlClassNameProvider>
    </DocProvider>;
}
