// docusaurus-theme-openapi-docs/lib/theme/ApiDemoPanel/FormFileUpload/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _FloatingButton = _interopRequireDefault(require("@theme/ApiDemoPanel/FloatingButton"));
var _reactMagicDropzone = _interopRequireDefault(require("react-magic-dropzone"));
var _stylesModule = _interopRequireDefault(require("./styles.module.css"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

function RenderPreview({
  file
}) {
  switch (file.type) {
    case "image/png":
    case "image/jpeg":
    case "image/jpg":
    case "image/svg+xml":
      return <img style={{
        borderRadius: "4px"
      }} src={file.preview} alt="" />;
    default:
      return <div style={{
        display: "flex",
        alignItems: "center",
        minWidth: 0
      }}>
          <svg viewBox="0 0 100 120" style={{
          width: "50px",
          height: "60px"
        }}>
            <path fillRule="evenodd" fill="#b3beca" d="M100.000,39.790 L100.000,105.000 C100.000,113.284 93.284,120.000 85.000,120.000 L15.000,120.000 C6.716,120.000 -0.000,113.284 -0.000,105.000 L-0.000,15.000 C-0.000,6.716 6.716,-0.000 15.000,-0.000 L60.210,-0.000 L100.000,39.790 Z" />
            <path fillRule="evenodd" fill="#90a1b1" transform="translate(60, 0)" d="M0.210,-0.000 L40.000,39.790 L40.000,40.000 L15.000,40.000 C6.716,40.000 0.000,33.284 0.000,25.000 L0.000,-0.000 L0.210,-0.000 Z" />
          </svg>
          <div className={_stylesModule.default.filename}>{file.name}</div>
        </div>;
  }
}
function FormFileUpload({
  placeholder,
  onChange
}) {
  const [hover, setHover] = (0, _react.useState)(false);
  const [file, setFile] = (0, _react.useState)();
  function setAndNotifyFile(file) {
    setFile(file);
    onChange === null || onChange === void 0 ? void 0 : onChange(file);
  }
  function handleDrop(accepted) {
    const [file] = accepted;
    setAndNotifyFile(file);
    setHover(false);
  }
  return <_FloatingButton.default>
      <_reactMagicDropzone.default className={hover ? _stylesModule.default.dropzoneHover : _stylesModule.default.dropzone} onDrop={handleDrop} onDragEnter={() => setHover(true)} onDragLeave={() => setHover(false)} multiple={false} style={{
      marginTop: "calc(var(--ifm-pre-padding) / 2)"
    }}>
        {file ? <>
            <button style={{
          marginTop: "calc(var(--ifm-pre-padding) / 2)"
        }} onClick={e => {
          e.stopPropagation();
          setAndNotifyFile(undefined);
        }}>
              Clear
            </button>
            <RenderPreview file={file} />
          </> : <div className={_stylesModule.default.dropzoneContent}>{placeholder}</div>}
      </_reactMagicDropzone.default>
    </_FloatingButton.default>;
}
var _default = FormFileUpload;
exports.default = _default;
