// docusaurus-theme-openapi-docs/lib/theme/ApiDemoPanel/CodeTabs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CodeTabs;
var _react = _interopRequireWildcard(require("react"));
var _themeCommon = require("@docusaurus/theme-common");
var _useIsBrowser = _interopRequireDefault(require("@docusaurus/useIsBrowser"));
var _Curl = require("@theme/ApiDemoPanel/Curl");
var _clsx = _interopRequireDefault(require("clsx"));
var _stylesModule = _interopRequireDefault(require("./styles.module.css"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
/* ============================================================================
 * Copyright (c) Palo Alto Networks
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 * ========================================================================== */

const {
  useScrollPositionBlocker,
  useTabGroupChoice
} = require("@docusaurus/theme-common/internal");

// A very rough duck type, but good enough to guard against mistakes while
// allowing customization
function isTabItem(comp) {
  return "value" in comp.props;
}
function TabsComponent(props) {
  var _ref, _children$find;
  const {
    lazy,
    block,
    defaultValue: defaultValueProp,
    values: valuesProp,
    groupId,
    className,
    action
  } = props;
  const children = _react.default.Children.map(props.children, child => {
    if ((0, _react.isValidElement)(child) && isTabItem(child)) {
      return child;
    }
    // child.type.name will give non-sensical values in prod because of
    // minification, but we assume it won't throw in prod.
    throw new Error(`Docusaurus error: Bad <Tabs> child <${
    // @ts-expect-error: guarding against unexpected cases
    typeof child.type === "string" ? child.type : child.type.name}>: all children of the <Tabs> component should be <TabItem>, and every <TabItem> should have a unique "value" prop.`);
  });
  const values = valuesProp !== null && valuesProp !== void 0 ? valuesProp :
  // Only pick keys that we recognize. MDX would inject some keys by default
  children.map(({
    props: {
      value,
      label,
      attributes
    }
  }) => ({
    value,
    label,
    attributes
  }));
  const dup = (0, _themeCommon.duplicates)(values, (a, b) => a.value === b.value);
  if (dup.length > 0) {
    throw new Error(`Docusaurus error: Duplicate values "${dup.map(a => a.value).join(", ")}" found in <Tabs>. Every value needs to be unique.`);
  }
  // When defaultValueProp is null, don't show a default tab
  const defaultValue = defaultValueProp === null ? defaultValueProp : (_ref = defaultValueProp !== null && defaultValueProp !== void 0 ? defaultValueProp : (_children$find = children.find(child => child.props.default)) === null || _children$find === void 0 ? void 0 : _children$find.props.value) !== null && _ref !== void 0 ? _ref : children[0].props.value;
  if (defaultValue !== null && !values.some(a => a.value === defaultValue)) {
    throw new Error(`Docusaurus error: The <Tabs> has a defaultValue "${defaultValue}" but none of its children has the corresponding value. Available values are: ${values.map(a => a.value).join(", ")}. If you intend to show no default tab, use defaultValue={null} instead.`);
  }
  const {
    tabGroupChoices,
    setTabGroupChoices
  } = useTabGroupChoice();
  const [selectedValue, setSelectedValue] = (0, _react.useState)(defaultValue);
  const tabRefs = [];
  const {
    blockElementScrollPositionUntilNextRender
  } = useScrollPositionBlocker();
  if (groupId != null) {
    const relevantTabGroupChoice = tabGroupChoices[groupId];
    if (relevantTabGroupChoice != null && relevantTabGroupChoice !== selectedValue && values.some(value => value.value === relevantTabGroupChoice)) {
      setSelectedValue(relevantTabGroupChoice);
    }
  }
  const handleTabChange = event => {
    const newTab = event.currentTarget;
    const newTabIndex = tabRefs.indexOf(newTab);
    const newTabValue = values[newTabIndex].value;
    if (newTabValue !== selectedValue) {
      blockElementScrollPositionUntilNextRender(newTab);
      setSelectedValue(newTabValue);
      if (action) {
        const newLanguage = _Curl.languageSet.filter(lang => lang.language === newTabValue);
        action(newLanguage[0]);
      }
      if (groupId != null) {
        setTabGroupChoices(groupId, String(newTabValue));
      }
    }
  };
  const handleKeydown = event => {
    var _focusElement;
    let focusElement = null;
    switch (event.key) {
      case "ArrowRight":
        {
          var _tabRefs$nextTab;
          const nextTab = tabRefs.indexOf(event.currentTarget) + 1;
          focusElement = (_tabRefs$nextTab = tabRefs[nextTab]) !== null && _tabRefs$nextTab !== void 0 ? _tabRefs$nextTab : tabRefs[0];
          break;
        }
      case "ArrowLeft":
        {
          var _tabRefs$prevTab;
          const prevTab = tabRefs.indexOf(event.currentTarget) - 1;
          focusElement = (_tabRefs$prevTab = tabRefs[prevTab]) !== null && _tabRefs$prevTab !== void 0 ? _tabRefs$prevTab : tabRefs[tabRefs.length - 1];
          break;
        }
      default:
        break;
    }
    (_focusElement = focusElement) === null || _focusElement === void 0 ? void 0 : _focusElement.focus();
  };
  return <div className={(0, _clsx.default)("tabs-container", _stylesModule.default.tabList)}>
      <ul role="tablist" aria-orientation="horizontal" className={(0, _clsx.default)("tabs", {
      "tabs--block": block
    }, _stylesModule.default.code__tabs, className)}>
        {values.map(({
        value,
        label,
        attributes
      }) => <li role="tab" tabIndex={selectedValue === value ? 0 : -1} aria-selected={selectedValue === value} key={value} ref={tabControl => tabRefs.push(tabControl)} onKeyDown={handleKeydown} onFocus={handleTabChange} onClick={handleTabChange} {...attributes} className={(0, _clsx.default)("tabs__item", _stylesModule.default.tabItem, attributes === null || attributes === void 0 ? void 0 : attributes.className, {
        "tabs__item--active": selectedValue === value
      })}>
            {label !== null && label !== void 0 ? label : value}
          </li>)}
      </ul>

      {lazy ? (0, _react.cloneElement)(children.filter(tabItem => tabItem.props.value === selectedValue)[0], {
      className: "margin-top--md"
    }) : <div className="margin-top--md">
          {children.map((tabItem, i) => (0, _react.cloneElement)(tabItem, {
        key: i,
        hidden: tabItem.props.value !== selectedValue
      }))}
        </div>}
    </div>;
}
function CodeTabs(props) {
  const isBrowser = (0, _useIsBrowser.default)();
  return <TabsComponent
  // Remount tabs after hydration
  // Temporary fix for https://github.com/facebook/docusaurus/issues/5653
  key={String(isBrowser)} {...props} />;
}
